### PR TITLE
fix: adding a first-class option for trusted proxies

### LINKED ISSUE #32135: Option to specify trusted proxies
### Description

Add a new `proxy-trusted-list` option (name is subject to discussion) for specifying a list of trusted IP addresses, hostnames or CIDR and should map to `quarkus.http.proxy.trusted-proxies` Quarkus option.

If not too involved, a new field should be added to the Keycloak CR too as part of this.

### Discussion

_No response_

### Motivation

For cases when Keycloak is exposed behind a reverse proxy, it should provide the capability to list addresses of such proxy. If a request is then NOT coming from any of the whitelisted addresses, it will ignore `Forwarded` and `X-Forwarded-*` headers.

### Details

Originally suggested by @ahus1.

### PR DESCRIPTION
**Pull request by shawkins**
closes: #32135

### CODE REVIEW

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for this PR. Some comments on the wordings around this feature.

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java
Comment (shawkins) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:63]: It's valid according to: https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus-http-proxy-trusted-proxies

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java
Comment (shawkins) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:68]: It's valid according to: https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus-http-proxy-trusted-proxies

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java
Comment (ahus1) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:63]: Thank you for pointing me at the docs. It seems that Quarkus actually implements it this way, still the documentation of the property in Keycloak is then incomplete as it doesn't point out that port numbers can be used with the trusted proxy address. Reading the Quarkus property again, this is then about "socket addresses", where I originally thought only host addresses would be specified.

Thinking about this a bit deeper, I think this is a strange configuration to also provide the port number: A TCP connection is identified by the the four elements source IP, source port, target port and target IP. The target port and target IP are where Quarkus listens for incoming connections. Source host is the proxy. If I also limit the source port, the proxy I configure here can only establish a single connection at a time. And even if that connection is closed, it might take a while until it can open a new connection while it is still in CLOSE_WAIT AFAIK. I'll raise a question with Quarkus, still waiting for the answer is out of scope for this PR as it is already implemented that way in Quarkus, and people might actually use it. 

I'll suggest a re-working in a minute.

File: quarkus/config-api/src/main/java/org/keycloak/config/ProxyOptions.java
Comment (ahus1) [inline quarkus/config-api/src/main/java/org/keycloak/config/ProxyOptions.java:63]:
```suggestion
            .description("A comma separated list of trusted proxy addresses. If set, then proxy headers from other addresses will be ignored. By default all addresses are trusted. A trusted proxy address should be specified as the IP address (IPv4 or IPv6) or Classless Inter-Domain Routing (CIDR) notation. Hostnames are discouraged as a DNS lookup is required. Optionally each address can be suffixed with a colon and a port number to restrict access even further.")
```

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java
Comment (shawkins) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:63]: Yes, I left it out of the description - it's already getting pretty long, I'm fine with making it longer as it seems that pointing directly to the quarkus docs would be frowned upon. As for what is in the tests, I just need a placeholder value that in the one case will cause the headers to be ignored. Those can of course be changed if you want.

Comment (ahus1) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:63]: Let's leave the placeholder as is, works for me. See below for one more sentence :-(

Comment (shawkins) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:63]: Looks good

Comment (ahus1) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyHostnameV2DistTest.java:63]: https://github.com/quarkusio/quarkus/issues/42760

**Review by ahus1 — DISMISSED** [review]
LGTM, thank you for this change. Still I think the help texts will fail with the latest change you accepted.

**Comment by shawkins** [comment]
> LGTM, thank you for this change. Still I think the help texts will fail with the latest change you accepted.

Corrected now.

**Comment by ahus1** [comment]
@shawkins - I'm testing how Quarkus is handling hostnames that require DNS resolving. I have a suspicion that it might be broken, so I put a "hold" on this one. Sorry for this. Will report back once I have some presentable results.

**Comment by ahus1** [comment]
I found that Quarkus is using `DnsClient.lookup()` to resolve the host name. According to https://github.com/eclipse-vertx/vert.x/issues/5035 this will work with some DNS servers, but not with others as it tries to resolve both `A` and `AAAA` records at the same time. I'll have to make up my mind one this mess. Will get back to you during the week.

**Comment by ahus1** [comment]
> Are you thinking that users, despite the recommendation, will still want to use hostnames for ease of use?

I'd say people will try to use it, and given the current description it sounds that it is still supported while not being recommended. 

I did some more analysis on how the forwarded header handling is doing the hostname handling, it IMHO it has several pitfalls: Sometimes DNS will not work, multiple A and AAAA records will not work, mixed IPv4 and IPv6 is problematic, and CNAMES won't work as well AFAIK, see https://github.com/quarkusio/quarkus/issues/42782 ... the Vert.x problem was only part of it.

Also the discussion around ports revealed that it is probably not working as intended when you add a port, see https://github.com/quarkusio/quarkus/issues/42760

So the only thing that works are IMHO either IP addresses or CIDRs. So I wonder if we could parse the input in Keycloak and sure that there are only those inputs and no ports or hostnames. 

I acknowledge that this is now bigger than what I originally asked for and anticipated. So it is ok if this moves to another sprint to get re-planned and discussed in the team first. 

Let me hear your thoughts.

**Comment by shawkins** [comment]
> So the only thing that works are IMHO either IP addresses or CIDRs. So I wonder if we could parse the input in Keycloak and sure that there are only those inputs and no ports or hostnames.

I'm perfectly fine with that restriction. It will greatly simplify our description.

**Comment by shawkins** [comment]
@ahus1 updated with additional validation.

**Review by ahus1 — CHANGES_REQUESTED** [review]
@shawkins - thank you for your quick turnaround. I agree with this approach. Still it seems some of the tests are now failing around feature flags, either due to changes messages or a changed logic. Please have a look. Once the tests pass, it should be good for merging.

**Comment by ahus1** [comment]
@shawkins - one part of the Operator test run failed. Don't think it is related, still strange: 

> Error details:
jakarta.enterprise.inject.CreationException: Error creating synthetic bean [Le6zQbzkojAYO_OiKIQWJf4lGa4]: jakarta.enterprise.inject.UnsatisfiedResolutionException: No bean found for required type [class io.quarkus.opentelemetry.runtime.tracing.DelayedAttributes] and qualifiers [[@jakarta.enterprise.inject.Any()]]

**Comment by shawkins** [comment]
> @shawkins - one part of the Operator test run failed. Don't think it is related, still strange:

Not sure what's going on there yet. It happened on both of my open prs, so it doesn't seem directly related to either.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ProxyPropertyMappers.java
Comment (vmuzikar) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ProxyPropertyMappers.java:61]: Do we need this? Quarkus doesn't perform any validations?

**Review by vmuzikar — COMMENTED** [review]
@ahus1 @shawkins Going through the discussion here, it looks like some valid concerns were raised. But I'm wondering if we should be really trying to address them at the Keycloak level. I am not a big fan of introducing additional fixes around Quarkus features we're exposing as it feels like an unnecessary maintenance burden. Should we rather raise this with Quarkus?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ProxyPropertyMappers.java
Comment (shawkins) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ProxyPropertyMappers.java:61]: Quarkus does perform validation, but based upon the issues raised and the general recommendation to not use hostnames we're restricting to a subset of what quarkus allows.

Comment (vmuzikar) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ProxyPropertyMappers.java:61]: It's then a question whether this is rather a Quarkus issue and as such should be raised with them. But I already added a comment about that above, so resolving this one.

**Comment by shawkins** [comment]
> @ahus1 @shawkins Going through the discussion here, it looks like some valid concerns were raised. But I'm wondering if we should be really trying to address them at the Keycloak level. I am not a big fan of introducing additional fixes around Quarkus features we're exposing as it feels like an unnecessary maintenance burden. Should we rather raise this with Quarkus?

@ahus1 has raised two issues. The concerns are: even if the dns issue is improved, it will still incur overhead vs addresses/CIDR so why expose that for use at this time. The current port restriction seems unusable in practice. So we have several choices:
- just document what quarkus accepts - that's where the pr was initially headed, but that just seems to be opening us up to additional support problems
- document restricted expectations and validate the config matches that - this is the current state of the pr
- document restricted expectations and let the config values flow to quarkus - is this your prefered option @vmuzikar ? Are you assuming that users won't try anything beyond what is documented, or at least if they do that we can simply say it's not supposed to be supported?

**Comment by ahus1** [comment]
@vmuzikar 

>  But I'm wondering if we should be really trying to address them at the Keycloak level. I am not a big fan of introducing additional fixes around Quarkus features we're exposing as it feels like an unnecessary maintenance burden.

I'm not considering this a workaround, rather than a filter on what we want to support in Keycloak given the functionality Quarkus provides. IMHO putting a hostname there which triggers a DNS resolve on every request is something that Quarkus considers a feature, but I'd never recommend or want to support that in Keycloak as it will be the source of slowdowns. Having the option to provide a port as a feature in Quarkus might be more of a misunderstanding in the first place, and it is IMHO difficult to get right given the current setup. 

I'd rather have this simpler option in Keycloak, than exposing the whole complexity Quarkus adds to it. 

So I'd say it should be merged as is.

**Review by ahus1 — DISMISSED** [review]
Thank you for this change. With the filtered functionality this should now be simple to use for Keycloak users.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:361]: What would be the execution order? 

IMO, the validator execution order should be:

1. default validator or specified in `validator()`
2. top-down builder `appendValidator()`

So, If I'm not mistaken, this might be preferred, right?
```suggestion
                current.accept(mapper, value);
                validator.accept(mapper, value);
```

**Review by mabartos — COMMENTED** [review]
@shawkins Just one thing, otherwise LGTM. 

I agree to have custom validators for the trusted proxies as we might prevent issues with the hostname resolution and the pragmatic non-usability of the single port, as @ahus1 nicely described in the Quarkus issues.

BTW we should think about the validation mechanism for properties and potentially unify it somehow for better UX - even with the `isEnabled()`. But it's out of the scope of this.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java
Comment (shawkins) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:361]: This change was leftover from my initial resolution of the validation issues that cropped up, but is not longer needed given the separate methods. 

I can address this in a follow up, since it doesn't come into play with this PR. What it should do is collect the validation errors.

**Review by vmuzikar — DISMISSED** [review]
Thinking about it more, while the port might be considered a bug on Quarkus side, I believe there's only a slim chance Quarkus would reconsider the support for hostnames. I agree with addressing that on Keycloak side as is.

@shawkins Thank you for the changes!

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java
Comment (mabartos) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:361]: Ok, agree to polish validations in a follow-up task. It's ok to provide these changes in this PR.

**Review by mabartos — DISMISSED** [review]
LGTM. 

Validation mechanism improvements will be done in a follow-up issue.

**Comment by vmuzikar** [comment]
@shawkins Could you please rebase?

**Review by vmuzikar — APPROVED** [review]
@shawkins Thanks for the rebase.

@ahus1 Since you are the assignee, do you want to give it another look? Otherwise, I believe we're ready to merge.

**Comment by ahus1** [comment]
I'm adding myself as an assignee to keep track of issues I care about :-) 

Great work everyone on this. Merging!

**Pull request merged**
